### PR TITLE
MDEV-40180 Cache constant args in JSON_EQUALS and JSON_OVERLAPS

### PR DESCRIPTION
**Pull request by VasuBhakt**
Fixes [MDEV-40180](https://jira.mariadb.org/browse/MDEV-40180) 

# Problem
`JSON_EQUALS` evaluates and parses constant arguments per-row. `JSON_OVERLAPS` caches constant arguments in `fix_length_and_dec`, which fails to recalculate during `--ps-protocol` executions with changing bound parameters.

# Solution
Migrated constant argument caching logic to `::val_bool()` for both functions:

1. **`Item_func_json_equals`**: Added deep-copy caching of the normalized `DYNAMIC_STRING` upon first evaluation of a constant argument. Subsequent evaluations bypass `json_normalize_engine` and reuse the cached string.
2. **`Item_func_json_overlaps`**: Removed flawed `a2_constant` caching from `fix_length_and_dec` and implemented per-execution caching of the evaluated string inside `val_bool()`.
3. **Prepared Statements**: Overrode `cleanup()` in both classes to reset boolean caching flags, ensuring correct cache invalidation between executions.

*(Note: The nested object short-circuit evaluation will be addressed in a follow-up incremental PR).*

# Tests
Executed the `json`, `json_equals`, and `json_normalize` MTR suites (both standard and `--ps-protocol`) to verify behavior preservation and resolution of the prepared statement cache bug.

# Version
The change is directed at versions `10.11`, `11.4`, `11.8`, and `12.3`. The PR is based on branch `10.11`, being the lowest affected branch.

### CODE REVIEW

File: sql/item_jsonfunc.cc [hunk 1]
```diff
@@ -622,48 +622,107 @@ bool Item_func_json_equals::val_bool()
...
   {
-    dynstr_free(&a_res);
-    null_value= 1;
-    return 1;
+    if(a_null) 
+      goto return_null;
   }
+  else 
+  {
+    a= args[0]->val_json(&a_tmp);
```
Comment (grooverdan) [inline sql/item_jsonfunc.cc:646]: note after next rebase the null checking of `a` needs to be after this. 63d779cff5321a4276f0d16ce05c5899b745099a by @vuvova corrected my mistake (thank you!), lets not regress it.

File: sql/item_jsonfunc.cc [hunk 2]
```diff
@@ -622,48 +622,107 @@ bool Item_func_json_equals::val_bool()
...
 
-  DYNAMIC_STRING b_res;
-  if (init_dynamic_string(&b_res, NULL, 0, 0))
+  /* Process First Argument */
+  if (a_const && a_parsed)
   {
-    dynstr_free(&a_res);
-    null_value= 1;
-    return 1;
+    if(a_null)
```
Comment (grooverdan) [inline sql/item_jsonfunc.cc:641]: space after `if`

File: sql/item_jsonfunc.cc [hunk 3]
```diff
@@ -622,48 +622,107 @@ bool Item_func_json_equals::val_bool()
...
   }
+  else 
+  {
+    a= args[0]->val_json(&a_tmp);
+    if (!a)
+      goto set_a_null;
+    
+    if (!cached_a.str)
+    {
+      if (init_dynamic_string(&cached_a, NULL, 0, 0))
```
Comment (grooverdan) [inline sql/item_jsonfunc.cc:652]: As this is going to contain a normalized form of `a`, use `a->length()` as the 3rd arg for the initial size.

The failure of this is a memory allocation failure. While it hasn't been done well in other examples in this file yet, the response is:
```
my_error(ER_OUTOFMEMORY, MYF(0), a->length());
goto return_null
```
As a pushed error it shouldn't return here

File: sql/item_jsonfunc.cc [hunk 4]
```diff
@@ -622,48 +622,107 @@ bool Item_func_json_equals::val_bool()
...
+        goto set_a_null;
+    }
+    else 
+    {
+      cached_a.length= 0;
```
Comment (grooverdan) [inline sql/item_jsonfunc.cc:657]: just a comment in code here about resetting string for next value.

File: sql/item_jsonfunc.cc [hunk 5]
```diff
@@ -622,48 +622,107 @@ bool Item_func_json_equals::val_bool()
...
-    return 1;
-  }
+  thd= current_thd;
+  JSON_DO_PAUSE_EXECUTION(thd, 0.0002);
+  je.killed_ptr= (uint32_t *) &thd->killed;
 
-  DYNAMIC_STRING b_res;
-  if (init_dynamic_string(&b_res, NULL, 0, 0))
+  /* Process First Argument */
+  if (a_const && a_parsed)
```
Comment (grooverdan) [inline sql/item_jsonfunc.cc:639]: `a_parsed` is always the same as `cached_a.str != nullptr` so I think `a_parsed` can be eliminated.

File: sql/item_jsonfunc.cc [hunk 6]
```diff
@@ -622,48 +622,107 @@ bool Item_func_json_equals::val_bool()
...
+    if (json_normalize_engine(&je, &cached_a, a->ptr(), a->length(), a->charset())) 
+    {
+      goto set_a_null;
+    }
 
-  if (json_normalize_engine(&je, &a_res, a->ptr(), a->length(), a->charset()))
-    goto return_null;
+    if (a_const)
+    {
+      a_null= false;
```
Comment (grooverdan) [inline sql/item_jsonfunc.cc:667]: I think can set `a_null` unconditionally. Its only looked at under `a_const ==true` anyway0.

File: sql/item_jsonfunc.cc [hunk 7]
```diff
@@ -622,48 +622,107 @@ bool Item_func_json_equals::val_bool()
...
+    }
+  }
 
   arg_num++;
-  if (json_normalize_engine(&je, &b_res, b->ptr(), b->length(), b->charset()))
-    goto return_null;
 
-  result= strcmp(a_res.str, b_res.str) ? 0 : 1;
+  /* Process Second Argument */
+  if (b_const && b_parsed)
```
Comment (grooverdan) [inline sql/item_jsonfunc.cc:675]: same comments on  first arg parsing apply here too.

File: sql/item_jsonfunc.cc [hunk 8]
```diff
@@ -5046,17 +5103,34 @@ bool Item_func_json_overlaps::val_bool()
...
-    val= args[1]->val_json(&tmp_val);
-    a2_parsed= a2_constant;
+    val= (cached_val.is_alloced() || cached_val.length()) ? &cached_val : 0;
+  }
+  else 
+  {
+    String *v= args[1]->val_json(&tmp_val);
+    if (v) 
+    {
+      cached_val.copy(v->ptr(), v->length(), v->charset());
```
Comment (grooverdan) [inline sql/item_jsonfunc.cc:5123]: still copying here.

```
if (b_const)
{
   if (cached_val.is_allocated())
      val= &cached_val; /* or something*/
   else
   {
      val= args[1]->val_json(&cached_val);
   }
}
else
{
  something using local val= tmp_str (as local)
}
```
(incomplete)

**Review by grooverdan — COMMENTED** [review]
didn't get time to look at overlaps properly.

with all faults of json_equals - needs test case of an actual table with null/nonnul and constant comparison like

https://github.com/MariaDB/server/pull/5420/changes#diff-1905b4d26db3e2b6344ebfcd668c479351e8ab90301e42a91da2bdbdf7d20afeR4405-R4410

File: sql/item_jsonfunc.cc [hunk 9]
```diff
@@ -622,48 +622,107 @@ bool Item_func_json_equals::val_bool()
...
   String a_tmp, b_tmp;
+  String *a= nullptr, *b= nullptr;
   THD *thd;
   json_engine_t je;
+  bool a_const= args[0]->const_item(), b_const= args[1]->const_item();
 
   if ((null_value= args[0]->null_value || args[1]->null_value))
     return 1;
 
-  String *a= args[0]->val_json(&a_tmp);
```
Comment (gkodinov) [inline sql/item_jsonfunc.cc:631]: This goes beyond the scope of the preliminary review, so consider it optional. But I wanted to air it out anyway: 

I believe that arguments to these functions broadly fall into the following categories:
1. an SQL constant
2. a predicate
3. something repeating coming from a table.

This is very similar to the regular expressions case IMHO. 

In case 1 you know that a constant is a constant and can pre-parse at "compile time" (e.g. at fix_length_and_dec as Item_func_regexp_match() does). 

Case 2 should be optimized differently to start with: index or something. Failing that, It is highly unlikely IMHO that arguments in this case would be grouped together (sorted on one of the arguments). So you'd end up just maintaining the cache and looking for a hit. And this might even make things worse speed-wise. And shouldn't be a target of optimization. If such an optimization is to be implemented, it needs to be implemented as a source transformation of the arguments during query compilation, e.g. using some specialized form of Item_cache that can hold the parsed JSON, if there isn't one already. 

Case 3 is so unlikely that I believe that detecting and optimizing that is just hard. 

Lazy caching at runtime is IMHO just making things worse: for actual constants you just add extra instructions per hit (is this the same string as what's cached, if yes, then reuse; if not cache) and this will make things worse compared to the approach taken by the regexp functions. 

Thus I strongly believe that caching in fix_length_and_dec() *is* the right phase to do this. But it needs to be done correctly wrt prepared statements. You're saying there's a bug in these. Maybe fix that instead?

**Review by gkodinov — CHANGES_REQUESTED** [review]
Thank you for your contribution. This is a preliminary review.
The idea is great. But I do not know if it's not a new feature or refactoring. I believe it is. And, as such, it needs to go to main and also receive proper testing in the process.

So, please re-base to main, given that you want to implement the feature for the two functions and not just fix the bug in prepared statements for the one that already has it. 

Also, I have something to say about the approach taken. Feel free to ignore it since this is preliminary review. But, if you like the logic, please consider it.

**Comment by gkodinov** [comment]
There has been no activity on this for more than 3 weeks. Converting to Draft. Please re-open when/if you resume working on it.
